File: 2019/src/main/kotlin/day04/day4.kt
package day04

fun split(st: String): Sequence<String> = sequence {
    var index = 0
    while (index < st.length) {
        val sequence = Regex("[" + st[index] + "]+").find(st.substring(index))?.value ?: ""
        yield(sequence)
        index += sequence.length
    }
}

fun ok(password: String, tripletsAllowed: Boolean): Boolean {
    if (password.zipWithNext().any { (a, b) -> a > b }) return false
    return split(password).any { it.length >= 2 && (tripletsAllowed || it.length == 2) }
}

fun solve(input: String, tripletsAllowed: Boolean): Int {
    val (start, end) = input.split("-").map { it.toInt() }
    return (start..end).count { ok(it.toString(), tripletsAllowed) }
}

fun evaluatorOne(input: String): Int = solve(input, true)
fun evaluatorTwo(input: String): Int = solve(input, false)

fun main() {
    val input = "145852-616942"
    println("Part One: ${evaluatorOne(input)}")
    println("Part Two: ${evaluatorTwo(input)}")
}

File: 2019/src/main/scala/day04/day4.scala
package day04

import scala.collection.immutable.Range.Inclusive

def split(str: String) = new Iterator[Int] {
    private var idx = 0

    override def hasNext = idx < str.length

    override def next(): Int = {
        val ch = str(idx)
        val len = str.substring(idx).takeWhile(_ == ch).length
        idx += len
        len
    }
}

def sortedString(str: String) = (str.init zip str.tail).forall { case (a, b) => a <= b }

def partOneOkay(password: String) = split(password).exists(_ >= 2)
def partTwoOkay(password: String) = split(password).exists(_ == 2)

def solver(numRange: Inclusive, okayString: String => Boolean): Int = {
    return numRange.map(_.toString).filter(sortedString).count(okayString)
}

def evaluatorOne(input: Inclusive): Int = solver(input, partOneOkay)
def evaluatorTwo(input: Inclusive): Int = solver(input, partTwoOkay)

def main(): Unit = {
    val inputLine = "145852-616942"
    val Array(start, end) = inputLine.split("-").map(_.toInt)
    println(s"Part One: ${evaluatorOne(start to end)}")
    println(s"Part Two: ${evaluatorTwo(start to end)}")
}